runtime/oti/FFIUpcallThunkGenHelpers.hpp

```diff
@@ -323,4 +362,93 @@ AFI(I_8 *instructionPtr, I_32 sourceRegister, I_32 immediateInteger)
+}
+
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is larger than 4096 bytes.
+ */
+static I_32
+generateLargeStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea, I_32 sourceRegister, I_32 destinationRegister)
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numFullSlots = structSize / MAX_MVC_COPY_LENGTH;
+	const I_32 residue = structSize % MAX_MVC_COPY_LENGTH;
+	const I_32 maxLargeCopies = MAX_MVC_DISPLACEMENT / MAX_MVC_COPY_LENGTH;
+
+	const I_32 loopResidue = numFullSlots % maxLargeCopies;
+
+	// loopInstructionsLength = 16x MVC(6) + 2x LAY (6) + 1x AFI (6)
+	const I_32 loopInstructionsLength = 114; // size in bytes
+	I_32 destinationDisplacementInt = -1;
+	I_32 sourceDisplacementInt = -1;
+	I_32 currMVCLength = -1;
+
+	// Load loop counter into R3
+	totalInstructionSize += IILF(instructionPtr + totalInstructionSize, R0, numFullSlots - loopResidue);
+
+	// Generate a loop that will copy 4096 bytes at a time (i.e. 16 MVC instructions). The loop concludes when there
+	// are less than 4096 bytes left to copy.
+	for (int i = 0; i < maxLargeCopies; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		destinationDisplacementInt = i * MAX_MVC_COPY_LENGTH;
+		currMVCLength = MAX_MVC_COPY_LENGTH - 1;
+		sourceDisplacementInt = i * MAX_MVC_COPY_LENGTH;
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, destinationDisplacementInt, currMVCLength, destinationRegister, sourceDisplacementInt, sourceRegister);
+	}
+	totalInstructionSize += LAY(instructionPtr + totalInstructionSize, destinationRegister, 0, destinationRegister, MAX_MVC_DISPLACEMENT);
+	totalInstructionSize += LAY(instructionPtr + totalInstructionSize, sourceRegister, 0, sourceRegister, MAX_MVC_DISPLACEMENT);
+	totalInstructionSize += AFI(instructionPtr + totalInstructionSize, R0, -maxLargeCopies);
+
+	const I_32 conditionMask = 7;
+	totalInstructionSize += BRCL(instructionPtr + totalInstructionSize, conditionMask, -(loopInstructionsLength/2));
+
+	// If numberOfBytesLeftToCopy > 256 && numberOfBytesLeftToCopy < 4096, then we generate the appropriate number of MVC instructions to copy them sequentially below.
+	if (loopResidue > 0) {
+		for (I_32 i = 0; i < loopResidue; i++) {
+			// MVC D1(L,B1),D2(B2)
+			destinationDisplacementInt = i * MAX_MVC_COPY_LENGTH;
+			currMVCLength = MAX_MVC_COPY_LENGTH - 1;
+			sourceDisplacementInt = i * MAX_MVC_COPY_LENGTH;
+			totalInstructionSize += MVC(instructionPtr + totalInstructionSize, destinationDisplacementInt, currMVCLength, destinationRegister, sourceDisplacementInt, sourceRegister);
+		}
+	}
+
+	// If numberOfBytesLeftToCopy < 256 then we generate an instruction here to copy those bytes over.
+	if (residue != 0) {
+		// MVC D1(L,B1),D2(B2)
+		destinationDisplacementInt = loopResidue > 0 ? loopResidue * MAX_MVC_COPY_LENGTH : (numFullSlots - maxLargeCopies) * MAX_MVC_COPY_LENGTH;
```

Comment:
using `(numFullSlots - maxLargeCopies)` to calculate the remaining displacement looks wrong to me. example: numFullSlots = 48 while maxLargeCopies is a constant of 16 ... you end up with a very big displacement not fitting MVC instr?